functions/rate-limiter.js

```diff
@@ -0,0 +1,241 @@
+'use strict';
+
+const { getFirestore, Timestamp } = require('firebase-admin/firestore');
+const functions = require('firebase-functions');
+
+/**
+ * @typedef {Object} RateLimits
+ * @property {number} attempts - Total number of attempts made
+ * @property {number} successful - Number of successfully delivered notifications
+ * @property {number} errors - Number of failed notification attempts
+ * @property {number} total - Total number of notifications (successful + errors)
+ * @property {number} maximum - Maximum notifications allowed per day
+ * @property {number} remaining - Remaining notifications allowed today
+ * @property {Date} resetsAt - When the rate limit resets
+ */
+
+/**
+ * @typedef {Object} RateLimitStatus
+ * @property {boolean} isRateLimited - Whether the token has exceeded the rate limit
+ * @property {boolean} shouldSendRateLimitNotification - Whether to send a rate limit notification
+ * @property {RateLimits} rateLimits - Current rate limit statistics
+ */
+
+/**
+ * @typedef {Object} RateLimitData
+ * @property {number} attemptsCount - Number of attempts
+ * @property {number} deliveredCount - Number of delivered notifications
+ * @property {number} errorCount - Number of errors
+ * @property {number} totalCount - Total notifications sent
+ * @property {FirebaseFirestore.Timestamp} expiresAt - When this record expires
+ */
+
+/**
+ * Manages rate limiting for push notifications on a per-token basis.
+ * Each instance is specific to a single push token and maintains its own state.
+ */
+class RateLimiter {
+  /**
+   * Creates a new RateLimiter instance for a specific push token.
+   *
+   * @param {string} token - The push notification token to rate limit
+   * @param {number} [maxNotificationsPerDay=500] - Maximum notifications allowed per day
+   * @param {boolean} [debug=false] - Whether to enable debug logging
+   */
+  constructor(token, maxNotificationsPerDay = 500, debug = false) {
+    this.token = token;
+    this.db = getFirestore();
```

Comment:
This does not change?
Can we do this outside of the class so we don't call this on each construction?

Reply:
I've changed this to be required to avoid duplicate defaults though.